src/CodegenType.php

```diff
@@ -40,11 +40,16 @@ public function setType(string $type): this {
     return $this;
   }
 
-  public function setShape(CodegenShape $codegen_shape): this {
+  public function setShape(mixed $codegen_shape): this {
     invariant(
       $this->type === null,
       "You can't set both the type and the shape.",
     );
+    invariant(
+      $codegen_shape instanceof ICodeBuilderRenderer && ($codegen_shape instanceof CodegenShape_FUTURE || $codegen_shape instanceof CodegenShape),
```

Comment:
I'm fine with this for 3.x, but need a better solution for master before 4.x happens. Possibly different setters